datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -242,28 +214,92 @@ impl ExternalSorter {
 
         debug!("Spilling sort data of ExternalSorter to disk whilst inserting");
 
-        let tracking_metrics = self
-            .metrics_set
-            .new_intermediate_tracking(self.partition_id, &self.runtime.memory_pool);
+        self.in_mem_sort().await?;
 
         let spillfile = self.runtime.disk_manager.create_tmp_file("Sorting")?;
-        let stream = in_mem_partial_sort(
-            &mut self.in_mem_batches,
-            self.schema.clone(),
-            &self.expr,
-            self.session_config.batch_size(),
-            tracking_metrics,
-            self.fetch,
-        );
-
-        spill_partial_sorted_stream(&mut stream?, spillfile.path(), self.schema.clone())
-            .await?;
+        let batches = std::mem::take(&mut self.in_mem_batches);
+        spill_sorted_batches(batches, spillfile.path(), self.schema.clone()).await?;
         self.reservation.free();
         let used = self.metrics.mem_used().set(0);
         self.metrics.record_spill(used);
         self.spills.push(spillfile);
         Ok(used)
     }
+
+    /// Sorts the in_mem_batches in place
+    async fn in_mem_sort(&mut self) -> Result<()> {
+        if self.in_mem_batches_sorted {
+            return Ok(());
+        }
+
+        let tracking_metrics = self
+            .metrics_set
+            .new_intermediate_tracking(self.partition_id, &self.runtime.memory_pool);
+
+        self.in_mem_batches = self
+            .in_mem_sort_stream(tracking_metrics)?
+            .try_collect()
+            .await?;
+
+        let size: usize = self
+            .in_mem_batches
+            .iter()
+            .map(|x| x.get_array_memory_size())
+            .sum();
+
+        self.metrics.mem_used().set(size);
+        self.reservation.resize(size);
+        self.in_mem_batches_sorted = true;
+        Ok(())
+    }
+
+    /// Consumes in_mem_batches returning a sorted stream
+    fn in_mem_sort_stream(
+        &mut self,
+        metrics: MemTrackingMetrics,
+    ) -> Result<SendableRecordBatchStream> {
+        assert_ne!(self.in_mem_batches.len(), 0);
+        if self.in_mem_batches.len() == 1 {
+            let batch = self.in_mem_batches.remove(0);
+            let stream =
+                sort_batch_stream(batch, self.expr.clone(), self.fetch, metrics)?;
+            self.in_mem_batches.clear();
+            return Ok(stream);
+        }
+
+        // If less than 1MB of in-memory data, concatenate and sort in place
+        //
+        // This is a very rough heuristic and likely could be refined further
+        if self.reservation.size() < 1048576 {
+            // Concatenate memory batches together and sort
+            let batch = concat_batches(&self.schema, &self.in_mem_batches)?;
+            self.in_mem_batches.clear();
+            return sort_batch_stream(batch, self.expr.clone(), self.fetch, metrics);
+        }
+
+        let streams = self
```

Comment:
Thinking about this a bit more, we might want to group together groups of < 1MB to reduce the fan-in :thinking: